Fixes for archive migration documentation. (#866)
o1-labs/docs2: docs/berkeley-upgrade/migrating-archive-database-to-berkeley.mdx

# Mainnet to Berkeley Archive Migration

Before you start the migration, please read [How to prepare for archive migration](/berkeley-upgrade/planning-archive-migration).

## Download migration applications

Migration applications are distributed as part of the archive migration Docker and Debian packages.

> You can choose packages as you wish. The following commands help you install Debian packages or Docker images.

Reviewer: ```suggestion
Choose packages appropriate for your environment. 

The following commands help you install Debian packages or Docker images.
```

--mainnet-archive-uri {mainnet_connection_string} --migrated-archive-uri {migrated_connection_string} --mainnet-blocks-bucket 
{bucket name} --network "mainnet"
```
where: 
`--network` - precomputed blocks prefix 

##### Phase 2: Replayer in migration mode run

```
 mina-replayer --migration-mode --archive-uri {migrated_connection_string}  --input-file replayer-checkpoint-XXXXX.json --checkpoint-interval 10000
```

where `replayer-checkpoint-XXXXX.json` - is last checkpoint generated from previous migration

Incremental migration can be run continuously on top initial migration or last incremental until fork block is anounced 

#### Stage 3: Remainder migration

When fork block is announced we can tackle the remainder migration this is the last migration run 
we would want to perform.

- Inputs
   - Newest mainnet database
   - Mainnet genesis ledger
   - Replayer checkpoint from last run 
   - Migrated berkeley schema from last run
   - Fork block state hash

- Outputs
   - Migrated mainnet database to berkeley from up fork point
   - Replayer checkpoint which can be used for replayer run on berkeley schema

##### Phase 1: Berkeley migration app run

> Noticed decrease in `batch-size` for precomputed blocks as tool may hit missing precomputed blocks problem if we would like 
> to fetch too many of them in single batch while trying to migrate newest blocks in the chain

Reviewer: ```suggestion
You might notice a decrease in `batch-size` for precomputed blocks as the tool might hit missing precomputed blocks problem if you fetch too many of them in a single batch while trying to migrate the newest blocks in the chain:
```

Reviewer: @dkijania this sentence introduces a command, so I guessed that a user might notice a decreased batch size, maybe instead it needs to say "to reduce the batch-size?"

```
mina-berkeley-migration --batch-size 2 --config-file genesis_ledgers/mainnet.json
--mainnet-archive-uri {mainnet_connection_string} --migrated-archive-uri {migrated_connection_string} --mainnet-blocks-bucket 
{bucket name} --network "mainnet" **--fork-state-hash {fork-state-hash}**
```
where: 
`--network` - precomputed blocks prefix 

> `--fork-state-hash` last migrated block which will be used for as a fork point 
> 
> :info: When running berkeley migration app with fork-state-hash, there is no requirement for fork state block to be canonical.
> Tool will automatically converts all pending blocks in subchain including fork block to canonical

Reviewer: ```suggestion
`--fork-state-hash` last migrated block which will be used as a fork point 

**Info** When running the **berkeley-migration** app with fork-state-hash, there is no requirement for the fork state block to be canonical.
The tool automatically converts all pending blocks in the subchain, including the fork block, to canonical blocks.
```

   ```

5. Stage 2: Initial migration

   5.a) Phase 1:

   ```sh
   mina-berkeley-migration.exe --batch-size 2000 --config-file /etc/mina/genesis_ledgers/mainnet.json --mainnet-archive-uri postgres://postgres:postgres@localhost/archive_balances_migrated --migrated-archive-uri postgres://postgres:postgres@localhost/berkeley_migrated --mainnet-blocks-bucket mina_network_block_data --network mainnet
   ```

   5.b) Phase 2:

   ```sh
   mina-replayer --migration-mode --archive-uri postgres://postgres:postgres@localhost/ --input-file checkpoint-XXXX.json --checkpoint-interval 100  --checkpoint-file-prefix migration
   ```

5. Stage 2: Remainder migration

   5.a) Phase 1:

   ```sh
   mina-berkeley-migration.exe --batch-size 2000 --config-file /etc/mina/genesis_ledgers/mainnet.json --mainnet-archive-uri postgres://postgres:postgres@localhost/archive_balances_migrated --migrated-archive-uri postgres://postgres:postgres@localhost/berkeley_migrated --mainnet-blocks-bucket mina_network_block_data --network mainnet --fork-state-hash "3NLdCBNrDseiDKvVj8rZ15k2oAUvx4XuCc8mzf6fL2CmqTJVVceM" 
   ```

>    :warning: 3NLdCBNrDseiDKvVj8rZ15k2oAUvx4XuCc8mzf6fL2CmqTJVVceM - is some random hash, please do not user it on actual migration but learn from MF what is the fork point

Reviewer: ```suggestion
   :warning: 3NLdCBNrDseiDKvVj8rZ15k2oAUvx4XuCc8mzf6fL2CmqTJVVceM - is an example random hash, please **do not user this hash** on the actual migration. Use the official hash as provided by Mina Foundation for the fork point.
```

>    5.b) Phase 2:

Reviewer: @dkijania  I applaud this level of detail. After you get the technical content suggestions in and all content approved, I can help with the enumeration for the steps.

}
```

#### I am receiving errors from gsutil regarding missing precomputed blocks

When migrating last chunk of blocks from mainnet, there can be an issue for runs with
large `batch-size` argument inputs. As an optimization gsutil is trying to fetch as `block-size` amount of precomputed blocks.
However, when precomputed blocks bucket has not contain enough (= `block-size` amount) precomputed blocks gsutil may panic stopping migration process.
Our recommendation is to restart process with smaller number of `batch-size`. If you are running migration just after fork block is announced it is best
to have `--block-size` argument set up to 5.

#### How to start replayer migration from latest checkpoint

> Instead of using Mainnet ledger as input config, provide the checkpoint for the `--input-config` parameter.

Reviewer: we'll need to use consistently the same Markdown syntax and capitalization for "Mainnet ledger"
